### PR TITLE
Feat: Dynamic cube count dispatch

### PR DESCRIPTION
**Pull request by ArthurBrussee**
### Checklist
- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Changes

I have some custom kernels where only the GPU knows how many valid items were processed. Further dispatches then only need to process these items. WebGPU and co habe a mechanism for this - execute_indirect. This PR makes CubeCount an enum where either a static amount of work is dispatched, or a dynamic amount based on a buffer.

In CUDA land, the more common approach is to have kernels launch other kernels, which is a more general method of achieving this effect. I'm not sure if CudaRC has a mechanism to do this from their API. Creating a dummy kernel that launches the real kernel with the right settings could work. For now, it just reads the data from the buffer which isn't optimall but at least correct.

The code does get a bit annoying as CubeCount is now generic over the server. I wisth rust had the option to make CubeCount::Fixed a full valid return type! To prevent these generics further leaking into everything I've split the CubeCount and CubeTask. 

Things would also be a bit easier if the ComputeServer knew about CubeCount, but I don't want to make cube a dependency of burn-compute. One option is to move the CubeCount to burn-compute and just name it DispatchCount. For now I've done the same associated type trick as is in place for CubeTask.

### Testing

Tests, ran my own project using dynamic dispatch.

### CODE REVIEW

File: crates/burn-cuda/src/compute/server.rs [hunk 1]
```diff
@@ -101,12 +108,29 @@ impl<MM: MemoryManagement<CudaStorage>> ComputeServer for CudaServer<MM> {
...
+
+        let count = match count {
+            CubeCount::Static(x, y, z) => (x, y, z),
+            // TODO: CUDA doesn't have an exact equivalen of dynamic dispatch. Instead, kernels are free to launch other kernels.
+            // One option is to create a dummy kernel with 1 thread that launches the real kernel with the dynamic dispatch settings.
+            // For now, just read the dispatch settings from the buffer.
+            CubeCount::Dynamic(binding) => {
+                let data = self.read_sync(binding);
+                let data: Vec<u32> = bytemuck::cast_slice(&data).to_vec();
+                (data[0], data[1], data[2])
```
Comment (nathanielsimard) [inline crates/burn-cuda/src/compute/server.rs:129]: Not sure we are forced to call `to_vec`

```rust
let data: &[u32] = bytemuck::cast_slice(&data);
assert!(data.len() == 3, "Dynamic cube count should contain 3 values");
(data[0], data[1], data[2])
```

File: crates/burn-compute/src/server.rs [hunk 2]
```diff
@@ -17,6 +17,8 @@ where
 {
     /// The kernel type defines the computation algorithms.
     type Kernel: Send;
+    /// The type for the number of workgroups to dispatch.
+    type DispatchCount: Send;
```
[the remark this answers is not in the document]
Comment (nathanielsimard) [reply crates/burn-compute/src/server.rs:21]: Or even more general, dispatch options, which can be none in some tests.

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Thanks for the PR. I think there are only minor things to fix before we can merge. 

Sometime we use work groups and sometimes we use cube, we should be consistent and use cube most of the time (work group is WebGPU, block is CUDA)

File: crates/burn-cuda/src/compute/server.rs (hunk 1 above)
Comment (ArthurBrussee) [reply crates/burn-cuda/src/compute/server.rs]: Oh right nice spot!

**Comment by ArthurBrussee** [comment]
Tried to fix most of the workgroup -> cube conversions :) The CUDA compiler uses it quite a lot too, but leaving that for now.

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**